repository/src/main/java/org/apache/atlas/tasks/TaskQueueWatcher.java

```diff
@@ -88,41 +87,66 @@ public void run() {
 
                 List<AtlasTask> tasks = fetcher.getTasks();
                 if (CollectionUtils.isNotEmpty(tasks)) {
-                    addAll(tasks);
+                    final CountDownLatch latch = new CountDownLatch(tasks.size());
+                    submitAll(tasks, latch);
+                    waitForTasksToComplete(latch);
                 } else {
                     LOG.info("No tasks to queue, sleeping for {} ms", pollInterval);
+                    releaseLock(lock);
                 }
-
                 Thread.sleep(pollInterval);
-
             } catch (InterruptedException interruptedException) {
                 LOG.error("TaskQueueWatcher: Interrupted: thread is terminated, new tasks will not be loaded into the queue until next restart");
                 break;
             } catch (Exception e){
-                LOG.error("TaskQueueWatcher: Exception occurred");
-                throw e;
+                LOG.error("TaskQueueWatcher: Exception occurred " +e.getMessage(),e);
+            }
+            finally {
+                releaseLock(lock);
             }
         }
     }
 
-    private void addAll(List<AtlasTask> tasks) {
+    private InterProcessMutex acquireDistributedLock() throws Exception {
+        InterProcessMutex lockForTask = curatorFactory.lockInstanceWithDefaultZkRoot(TASK_LOCK);
+        LOG.info("Attempting to acquire a lock on task");
+        lockForTask.acquire();
+        LOG.info("Acquired a lock on task");
+        return lockForTask;
+    }
+
+    private void waitForTasksToComplete(final CountDownLatch latch) throws InterruptedException {
+        if (latch.getCount() != 0) {
+            LOG.info("TaskQueueWatcher: Waiting on Latch, current count: {}", latch.getCount());
+            latch.await();
+            LOG.info("TaskQueueWatcher: Waiting completed on Latch, current count: {}", latch.getCount());
+        }
+    }
+
+    private void releaseLock(InterProcessMutex lock) {
+        try {
+            if (isActiveActiveHAEnabled && lock != null && lock.isOwnedByCurrentThread()) {
+                LOG.info("About to release task lock");
+                lock.release();
+                LOG.info("successfully released task lock");
+            }
+        } catch (Exception e) {
+            LOG.error(e.getMessage(), e);
```

Comment:
Let's improve all the error logs by adding a custom message, so we can search on sentry.

Reply:
This is addressed